### PR TITLE
Native VLAN based IP Forwarding for cloud (EXPERIMENTAL) 

### PR DESCRIPTION
**Pull request by abhinandanpb**
The Changes include : 
1) Vlan based native ip forward for non bgp based L3 routing.
2) Bgp cleanup and fixes 

The vlan based native IP forwarding will use custom contiv control plane to distribute endpoint information and act as the routing node. The feature is still experimental. 
 
Advantages:
1) Container endpoint macs will not be visible on physical switches (solves mac scaling issue)
2) The Physical switches need not have bgp control plane as a prerequisite.
3) No VXLAN based tunneling since its vlan encap.
4) Co-exits with BGP control plane.

### CODE REVIEW

File: vlrouter.go [hunk 1]
```diff
@@ -64,18 +64,20 @@ type Vlrouter struct {
...
 	myBgpPeer     string             // bgp neighbor
 	unresolvedEPs cmap.ConcurrentMap // unresolved endpoint map
+	uplinkOfp     uint32             // uplink intf portno mapping
+
 }
 
 // GetUplink API gets the uplink port with uplinkID from uplink DB
-func (self *Vlrouter) GetUplink(uplinkID string) *PortInfo {
-	uplink, ok := self.uplinkPortDb.Get(uplinkID)
+func (vl *Vlrouter) GetUplink(uplinkID string) *PortInfo {
```
Comment (DivyaVavili) [inline vlrouter.go]: Minor nit: can we skip this change or make a separate commit in the PR? It's taking the focus off the actual changes in the PR.

File: vlrouter.go [hunk 2]
```diff
@@ -640,22 +690,36 @@ func (self *Vlrouter) AddEndpoint(endpoint *OfnetEndpoint) error {
...
-		self.unresolvedEPs.Remove(endpoint.EndpointID)
-		return nil
+	flowId := endpoint.EndpointID
+
+	if vl.agent.isExternalBgp(endpoint) {
+		flowId = flowId + "external"
+		vl.myBgpPeer = ""
 	}
 
+	//Delete the endpoint if it is in the cache
```
Comment (DivyaVavili) [inline vlrouter.go]: please remove if not required

File: vlrouter.go [hunk 3]
```diff
@@ -900,81 +969,100 @@ func (self *Vlrouter) initFgraph() error {
...
+			endpoint := vl.agent.getEndpointByIpVrf(arpHdr.IPDst, "default")
 			if endpoint == nil {
 				// Look for a service entry for the target IP
-				proxyMac := self.svcProxy.GetSvcProxyMAC(arpHdr.IPDst)
+				proxyMac := vl.svcProxy.GetSvcProxyMAC(arpHdr.IPDst)
 				if proxyMac == "" {
 					// If we dont know the IP address, dont send an ARP response
-					log.Debugf("Received ARP request for unknown IP: %v", arpHdr.IPDst)
-					self.agent.incrStats("ArpReqUnknownDest")
+					log.Infof("Received ARP request for unknown IP: %v", arpHdr.IPDst)
```
Comment (DivyaVavili) [inline vlrouter.go]: Retain as Debugf

File: vlrouter.go [hunk 4]
```diff
@@ -900,81 +969,100 @@ func (self *Vlrouter) initFgraph() error {
...
+							intf, err = net.InterfaceByName(uplink.MbrLinks[0].Name)
+							if err != nil {
+								log.Errorf("Error getting interface information. Err: %+v", err)
+								return
+							}
+						} else if vl.uplinkPortDb.Count() > 0 {
+							for ul := range vl.uplinkPortDb.IterBuffered() {
+								uplink := ul.Val.(*PortInfo)
+								if uplink != nil && len(uplink.MbrLinks) > 0 {
+									intf, err = net.InterfaceByName(uplink.MbrLinks[0].Name)
```
Comment (DivyaVavili) [inline vlrouter.go]: Can you handle the case where we have more than one uplink? I understand that we allow only one uplink as of now, but it would be good to handle all the scenarios in cases like these. It would save a lot of debugging time when the support is added later.

File: ofnetBgp.go [hunk 5]
```diff
@@ -263,22 +257,14 @@ func (self *OfnetBgp) DeleteProtoNeighbor() error {
...
-			if ep.EndpointType == "internal" {
-				ep.PortNo = 0
-				self.agent.endpointDb.Set(ep.EndpointID, ep)
-				//We readd unresolved endpoints that were learnt via
-				//etcd
-				self.agent.datapath.AddEndpoint(ep)
-			} else if ep.EndpointType == "external" {
-				self.agent.endpointDb.Remove(ep.EndpointID)
-			}
-		}
```
Comment (abhinandanpb) [inline ofnetBgp.go]: bgp doesnt need to know about uplinks anymore.

File: ofnetBgp.go [hunk 6]
```diff
@@ -61,64 +63,45 @@ type OfnetBgpInspect struct {
...
 3) Kicks off routines to monitor route updates and peer state
 */
 func (self *OfnetBgp) StartProtoServer(routerInfo *OfnetProtoRouterInfo) error {
 
-	log.Infof("Starting the Bgp Server with %v", routerInfo)
-	//go routine to start gobgp server
+	log.Infof("Starting the Bgp Server with %#v", routerInfo)
+
+	//Flush previous external endpoints if any
+	self.agent.FlushEndpoints(OFNET_EXTERNAL)
```
Comment (abhinandanpb) [inline ofnetBgp.go]: on netplugin restart to flush out bgp endpoints

File: ofnetAgent.go [hunk 7]
```diff
@@ -1192,3 +1211,74 @@ func (self *OfnetAgent) getvlanVrf(vlan uint16) *string {
 func (self *OfnetAgent) AddNameServer(ns NameServer) {
 	self.nameServer = ns
 }
+
+func (self *OfnetAgent) isInternal(endpoint *OfnetEndpoint) bool {
```
Comment (abhinandanpb) [inline ofnetAgent.go]: will check it if its simplifies

File: ofnetAgent.go [hunk 8]
```diff
@@ -817,7 +827,15 @@ func (self *OfnetAgent) RemoveNetwork(vlanId uint16, vni uint32, Gw string, Vrf
...
 func (self *OfnetAgent) AddUplink(uplinkPort *PortInfo) error {
 	// Call the datapath
-	return self.datapath.AddUplink(uplinkPort)
+	err := self.datapath.AddUplink(uplinkPort)
+	if err != nil {
+		return err
+	}
+
+	if self.protopath != nil {
+		self.protopath.SetRouterInfo(uplinkPort)
```
Comment (abhinandanpb) [inline ofnetAgent.go]: do we need it to be a fatal error ? The check at this point is only safety check. Netplugin rightly has the required checks and it wont come to this at all.

File: ofnetAgent.go [hunk 9]
```diff
@@ -139,7 +140,15 @@ func NewOfnetAgent(bridgeName string, dpName string, localIp net.IP, rpcPort uin
 	agent.MyAddr = localIp.String()
 	agent.dpName = dpName
 	agent.arpMode = ArpProxy
-
+	if len(uplinkInfo) > 0 {
+		intf, err := net.InterfaceByName(uplinkInfo[0])
+		if err == nil {
+			agent.localMac = intf.HardwareAddr.String()
```
Comment (abhinandanpb) [inline ofnetAgent.go]: not needed. We only need one interface mac even if there are multiple uplinks.

File: ofnetAgent.go [hunk 10]
```diff
@@ -130,7 +131,7 @@ const (
 // Create a new Ofnet agent and initialize it
 func NewOfnetAgent(bridgeName string, dpName string, localIp net.IP, rpcPort uint16,
 	ovsPort uint16, uplinkInfo []string) (*OfnetAgent, error) {
-	log.Infof("Creating new ofnet agent for %s,%s,%d,%d,%d,%v \n", bridgeName, dpName, localIp, rpcPort, ovsPort, uplinkInfo)
+	log.Infof("Creating new ofnet agent for %s,%s,%d,%d,%d,%v \n", bridgeName, dpName, localIp, rpcPort, ovsPort)
```
Comment (abhinandanpb) [inline ofnetAgent.go]: will make the change

File: ofnetBgp.go [hunk 11]
```diff
@@ -130,18 +113,26 @@ func (self *OfnetBgp) StartProtoServer(routerInfo *OfnetProtoRouterInfo) error {
...
 	}
+
 	self.cc = conn
-	log.Debugf("Creating the loopback port ")
-	err = self.agent.ovsDriver.CreatePort(self.intfName, "internal", 1)
-	if err != nil {
-		log.Errorf("Error creating the port: %v", err)
+	link, err := netlink.LinkByName(self.intfName)
+	if err == nil {
+		netlink.LinkSetDown(link)
```
Comment (abhinandanpb) [inline ofnetBgp.go]: This is not uplink flaps. This is internal interface link flap usually needed to clear out arps/exisiting l2 neighborship in the kernel. Typical scenrios are bgp config updates.

File: ofnetAgent.go (hunk 8 above)
Comment (DivyaVavili) [inline ofnetAgent.go]: Well, fatal error would mean that we will run into issues when the fwd-mode change is applied. So, a warning or any other notification to user to mention that we are using a specific member link for uplink should be good enough.

File: ofnetBgp.go [hunk 12]
```diff
@@ -625,3 +592,13 @@ func (self *OfnetBgp) InspectProto() (interface{}, error) {
 
 	return OfnetBgpInspect, nil
 }
+
+func (self *OfnetBgp) SetRouterInfo(uplink *PortInfo) {
+	log.Infof("Received Set Router info for %v \n", uplink)
+	if uplink != nil && len(uplink.MbrLinks) != 1 {
+		log.Errorf("OfnetBgp currently supports only one uplink interface. Num uplinks active: %d", len(uplink.ActiveLinks))
+		return
```
Comment (DivyaVavili) [inline ofnetBgp.go]: Rethinking this scenario and as mentioned in the earlier comment in AddUplink(), we should not return an error and handle this situation and pick one of the uplink member links.

Comment (DivyaVavili) [inline ofnetBgp.go]: @gkvijay @abhinandanpb: FYI, this was the failure scenario we had discussed when fwd-mode change is triggered. From our discussion, we should be handling this by picking one of the member links.

File: vlrouter.go [hunk 13]
```diff
@@ -99,39 +101,39 @@ func NewVlrouter(agent *OfnetAgent, rpcServ *rpc.Server) *Vlrouter {
...
 
-// Handle new master added event
-func (self *Vlrouter) MasterAdded(master *OfnetNode) error {
+// MasterAdded handles new master added event
+func (vl *Vlrouter) MasterAdded(master *OfnetNode) error {
 
 	return nil
 }
 
 // Handle switch connected notification
```
Comment (abhinandanpb) [inline vlrouter.go]: update to ?

File: vlrouter.go (hunk 2 above)
Comment (abhinandanpb) [inline vlrouter.go]: This is place holder once static router commits go in.

File: vlrouter.go (hunk 3 above)
Comment (abhinandanpb) [inline vlrouter.go]: will do

File: vlrouter.go (hunk 4 above)
Comment (abhinandanpb) [inline vlrouter.go]: any reason this needs to be handled for multiple uplinks ? We need to always use 1 mac address. Its better to stick with the same mac. Its has no relation to multiple uplinks or 1 uplink.

File: vlrouter.go (hunk 13 above)
Comment (DivyaVavili) [inline vlrouter.go]: Just to comply with gofmt (function_name function_purpose). If gofmt doesn't complain, feel free to ignore it for now.

Comment (abhinandanpb) [inline vlrouter.go]: isnt it in that format :) ? It wasnt before

File: vlrouter.go (hunk 1 above)
Comment (abhinandanpb) [inline vlrouter.go]: i would ignore this change. Its just minor clean up to change self to vl to be more go compliant.

File: ofnetBgp.go (hunk 12 above)
Comment (abhinandanpb) [inline ofnetBgp.go]: This needs to be handled in netplugin and it is currently. This is just additional safety net.

File: vlrouter.go (hunk 13 above)
Comment (DivyaVavili) [inline vlrouter.go]: I guess you are looking at MasterAdded function. I'm referring to the function below(SwitchConnected) and others :)

File: vlrouter.go (hunk 4 above)
Comment (DivyaVavili) [inline vlrouter.go]: My bad, I was thinking it was a loop over Mbrlinks.

File: ofnetBgp.go (hunk 12 above)
Comment (DivyaVavili) [inline ofnetBgp.go]: As I understand, this was being handled in NewOfnetBgp() which ensures that only one member link was considered. Now that the check is removed in this PR, we have no other checks. Am I missing anything here?

Comment (abhinandanpb) [inline ofnetBgp.go]: https://github.com/contiv/netplugin/blob/master/netplugin/agent/state_event.go#L233

Comment (DivyaVavili) [inline ofnetBgp.go]: OK. This a new change I was not aware of. Thanks for the info.

But still, this seems to be triggered only during fwd-mode change. When the netplugin comes up for the first time, don't we still need to handle the scenario? Or, does netplugin start off with bridge mode and then switch to routing mode instantaneously (in which case we should be fine)?

**Comment by jojimt** [comment]
@abhinandanpb Can you please rebase?

**Comment by abhinandanpb** [comment]
@DivyaVavili , @jojimt the inherent assumption of ofnet is to error out if there are multiple uplinks in the existing code -> https://github.com/contiv/ofnet/blob/master/vlrouter.go#L1063
This would have ensured that the multiple uplink addition would be fatal and the code would have exited before setting the router info. So it was a no-op. There is also a unit test to verify if the vlrouter instantiation fails if there are multiple uplinks.
However I am changing the logic in ofnet now to non fatal - meaning we will assume the first uplink as the active uplink all over ofnet and not error-out unless there is no uplink members.  This can later be displayed in inspect for debugging purposes. PTAL

**Comment by DivyaVavili** [comment]
LGTM
